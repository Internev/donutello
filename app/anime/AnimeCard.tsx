import { GetAnimeListQuery, useGetAnimeListQuery, Media } from "@/graphql/generated/graphql"
import { Button, Card, CardBody, Heading, Image, Stack, Text } from "@chakra-ui/react"

type AnimeMedia = NonNullable<NonNullable<GetAnimeListQuery['Page']>['media']>[number]

const AnimeCard: React.FC<{ anime: any }> = ({ anime }) => {
  return (
    // <Card.Root width={180} height={360}>
    //   <Image
    //     src={anime?.coverImage?.large || ''}
    //     alt={`${anime?.title?.userPreferred} key image`}
    //     rounded="md"
    //     height="180px"
    //   />
    //   <Card.Body gap="2">
    //     <Card.Title>{anime?.title?.userPreferred}</Card.Title>
    //     <Card.Description>
    //       {anime?.type}
    //     </Card.Description>
    //     {/* <Text textStyle="2xl" fontWeight="medium" letterSpacing="tight" mt="2">
    //       $450
    //     </Text> */}
    //   </Card.Body>
    //   <Card.Footer gap="2">
    //     {/* <Button variant="solid">Buy now</Button>
    //     <Button variant="ghost">Add to cart</Button> */}
    //     Footer
    //   </Card.Footer>
    // </Card.Root>
    <Card maxW='sm'>
      <CardBody>
        <Image
          src={anime?.coverImage?.large || ''}
          alt={`${anime?.title?.userPreferred} key image`}
          borderRadius='lg'
        />
        <Stack mt='6' spacing='3'>
          <Heading size='md'>{anime?.title?.userPreferred}</Heading>
          <Text>
            This sofa is perfect for modern tropical spaces, baroque inspired
            spaces, earthy toned spaces and for people who love a chic design with a
            sprinkle of vintage design.
          </Text>
          <Text color='blue.600' fontSize='2xl'>
            $450
          </Text>
        </Stack>
      </CardBody>
      {/* <Divider /> */}
      {/* <CardFooter>
      <ButtonGroup spacing='2'>
        <Button variant='solid' colorScheme='blue'>
          Buy now
        </Button>
        <Button variant='ghost' colorScheme='blue'>
          Add to cart
        </Button>
      </ButtonGroup>
    </CardFooter> */}
    </Card>
  )
}

export default AnimeCard


