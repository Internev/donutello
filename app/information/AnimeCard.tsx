import { GetAnimeListQuery, useGetAnimeListQuery, Media } from "@/graphql/generated/graphql"
import { Button, Card, CardBody, CardFooter, Heading, Image, Popover, PopoverContent, PopoverTrigger, Show, Stack, Text } from "@chakra-ui/react"

type AnimeMedia = NonNullable<NonNullable<GetAnimeListQuery['Page']>['media']>[number]

const makeNiceDate = (startYear: string, endYear: string) => {
  if (startYear === endYear) {
    return startYear
  }
  return `${startYear} - ${endYear}`
}

const makeNiceRating = (rating: number) => {
  const color = rating >= 75 ? 'green.500' : rating >= 50 ? 'yellow.500' : 'red.500'
  return <Text color={color}>Average Rating: {rating}%</Text>
}

const AnimeCard: React.FC<{ anime: any }> = ({ anime }) => { // TODO: max sure I turn types back on for deployment!
  const showDate = makeNiceDate(anime.startDate?.year || '', anime.endDate?.year || '')

  const CardDisplay = () => {
    return (
      <CardBody>
        <Image
          height={'320px'}
          src={anime?.coverImage?.large || ''}
          alt='Green double couch with wooden legs'
          borderRadius='lg'
        />
        <Stack mt='2' spacing='3'>
          <Text>
            {anime?.title.userPreferred.length > 50 ? `${anime?.title.userPreferred.substring(0, 47)}...` : anime?.title.userPreferred}
          </Text>
        </Stack>
      </CardBody>
    )
  }

  return (
    <Card maxW='sm' height={'400px'} variant={'unstyled'}>
      <Show above="md">
        <Popover
          trigger="hover"
          placement="right-start"
          gutter={4}
        >
          <PopoverTrigger>
            <CardDisplay />
          </PopoverTrigger>
          <PopoverContent>
            <Stack spacing='3' p={3}>
              <Text>{anime?.type}</Text>
              {makeNiceRating(anime?.averageScore)}
              <Text>Favourites: {anime?.favourites}</Text>
              <Text>{showDate}</Text>
              <Button variant="ghost">View Details</Button>
            </Stack>
          </PopoverContent>
        </Popover>
      </Show>
      <Show below="md">
        <CardDisplay />
      </Show>
    </Card >
  )
}

export default AnimeCard
