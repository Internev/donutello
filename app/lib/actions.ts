// app/lib/actions.ts
'use server'

import { signIn, signOut, auth } from '@/auth'
import { AuthError } from 'next-auth'
import { redirect } from 'next/navigation'
import { z } from 'zod'
import { sql } from '@vercel/postgres'
import { revalidatePath } from 'next/cache'

const FormSchema = z.object({
  username: z.string().min(1, 'Username is required'),
  jobtitle: z.string().min(1, 'Job title is required'),
})

export async function authenticate(prevState: string | undefined, formData: FormData) {
  try {
    const validatedFields = FormSchema.safeParse({
      username: formData.get('username'),
      jobtitle: formData.get('jobtitle'),
    })

    console.log('validated fields:', validatedFields)
    if (!validatedFields.success) {
      return 'Invalid fields. Please check your input.'
    }

    const { username, jobtitle } = validatedFields.data

    await signIn('credentials', {
      username,
      jobtitle,
    })

    return undefined
  } catch (error) {
    if (error instanceof AuthError) {
      return 'Authentication failed.'
    }
    throw error
  }
}

export async function handleSignOut() {
  await signOut()
  redirect('/login')
}

const SettingsSchema = z.object({
  id: z.string(),
  username: z.string().min(1, 'Username must exist'),
  jobtitle: z.string().min(1, 'Job title must exist'),
})

export async function updateSettings(prevState: string | undefined, formData: FormData) {
  try {
    // Get current session to verify user
    const session = await auth()
    if (!session?.user) {
      return {
        message: 'You must be logged in to update settings',
        status: 'error',
      }
    }

    const validatedFields = SettingsSchema.safeParse({
      id: formData.get('id'),
      username: formData.get('username'),
      jobtitle: formData.get('jobtitle'),
    })

    if (!validatedFields.success) {
      return {
        message: 'Invalid fields. Please check your input.',
        status: 'error',
      }
    }

    const { id, username, jobtitle } = validatedFields.data

    // Verify the user is updating their own settings
    if (id !== session.user.id) {
      return {
        message: 'Unauthorized',
        status: 'error',
      }
    }

    // Update database
    await sql`
      UPDATE users
      SET username = ${username}, jobtitle = ${jobtitle}
      WHERE id = ${id}
    `

    // Update session by signing in again with new credentials
    await signIn('credentials', {
      username,
      jobtitle,
      redirect: false,
    })

    revalidatePath('/settings')

    return {
      message: 'Settings updated successfully',
      status: 'success',
    }
  } catch (error) {
    console.error('Failed to update settings:', error)
    return {
      message: 'Failed to update settings',
      status: 'error',
    }
  }
}