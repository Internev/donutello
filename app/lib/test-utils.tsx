import { ChakraProvider } from '@chakra-ui/react'
import { render } from '@testing-library/react'

import { ReactNode } from 'react'

const AllProviders = ({ children }: { children: ReactNode }) => (
  <ChakraProvider>{children}</ChakraProvider>
)

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }