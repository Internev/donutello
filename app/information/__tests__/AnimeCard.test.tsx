
import { render, screen } from '../../lib/test-utils'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import AnimeCard, { AnimeMedia } from '../AnimeCard'

// Mock data
const mockAnime = {
  id: 1,
  title: {
    userPreferred: 'Test Anime Title'
  },
  coverImage: {
    large: 'https://example.com/image.jpg'
  },
  startDate: {
    year: '2023'
  },
  endDate: {
    year: '2024'
  },
  type: 'Anime',
  averageScore: 85,
  favourites: 6000
} as unknown as AnimeMedia

const mockAnimeUnpopular = {
  ...mockAnime,
  favourites: 1000
} as unknown as AnimeMedia

describe('AnimeCard', () => {
  it('renders anime card with correct title', () => {
    render(<AnimeCard anime={mockAnime} />)
    expect(screen.getByText('Test Anime Title')).toBeInTheDocument()
  })

  it('renders popular anime with star icon', () => {
    render(<AnimeCard anime={mockAnime} />)
    expect(screen.getByText('⭐️')).toBeInTheDocument()
  })

  it('does not render star icon for unpopular anime', () => {
    render(<AnimeCard anime={mockAnimeUnpopular} />)
    expect(screen.queryByText('⭐️')).not.toBeInTheDocument()
  })

  it('shows additional information on hover', async () => {
    render(<AnimeCard anime={mockAnime} />)

    // Find the card body and hover over it
    const cardBody = screen.getByRole('group')
    await userEvent.hover(cardBody)

    // Check if hover content appears
    expect(screen.getByText('TV')).toBeInTheDocument()
    expect(screen.getByText('Rating: 85%')).toBeInTheDocument()
    expect(screen.getByText('♥ 6,000')).toBeInTheDocument()
    expect(screen.getByText('2023 - 2024')).toBeInTheDocument()
  })

  it('renders with correct date format for same year', () => {
    const sameYearAnime = {
      ...mockAnime,
      startDate: { year: '2023' },
      endDate: { year: '2023' }
    } as unknown as AnimeMedia
    render(<AnimeCard anime={sameYearAnime} />)
    expect(screen.getByText('2023')).toBeInTheDocument()
  })

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(<AnimeCard anime={mockAnime} />)
    expect(container).toMatchSnapshot()
  })
})