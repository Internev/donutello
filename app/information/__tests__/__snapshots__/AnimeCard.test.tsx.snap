// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimeCard matches snapshot 1`] = `
<div>
  <div
    class="chakra-card css-1bto0a9"
  >
    <div
      class="chakra-card__body css-1e4fehn"
      role="group"
    >
      <div
        class="css-79elbk"
      >
        <img
          alt="Test Anime Title"
          class="chakra-image__placeholder css-y8l31t"
          src="https://via.placeholder.com/250x320?text=Loading..."
        />
        <div
          class="css-14ebbme"
        >
          <p
            class="chakra-text css-mcpu91"
          >
            ⭐️
          </p>
        </div>
        <div
          class="css-yjux51"
        >
          <div
            class="chakra-stack css-2xph3x"
          >
            <p
              class="chakra-text css-14p6uhh"
            >
              TV
            </p>
            <p
              class="chakra-text css-1ae45b4"
            >
              Rating: 
              85
              %
            </p>
            <p
              class="chakra-text css-1aakaxo"
            >
              ♥ 
              6,000
            </p>
            <p
              class="chakra-text css-zx7ia"
            >
              2023 - 2024
            </p>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-wxtwm2"
      >
        <p
          class="chakra-text css-n1oncg"
        >
          Test Anime Title
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
